presto-main/src/main/java/com/facebook/presto/server/remotetask/ContinuousBatchTaskStatusFetcher.java

```diff
@@ -0,0 +1,191 @@
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Consumer;
+
+import static com.facebook.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+
+public class ContinuousBatchTaskStatusFetcher
+{
+    public class Task
+    {
+        TaskId taskId;
+        StateMachine<TaskStatus> taskStatus;
+        Codec<Map<TaskId, TaskStatus>> taskListStatusCodec;
+    }
+
+    // TaskId -> WorkerId (String(URI(TaskStatus.getSelf() (get worker id)))
+    private final ConcurrentHashMap<TaskId, String> idWorkerMap;
+    // String(URI(TaskStatus.getSelf() (get worker id))) -> WorkerTaskStatusFetcher (List<Task>)
+    private final ConcurrentHashMap<String, WorkerTaskStatusFetcher> workerTaskMap;
+
+    private static final Logger log = Logger.get(ContinuousBatchTaskStatusFetcher.class);
+
+    private final Consumer<Throwable> onFail;
+    private final Duration refreshMaxWait;
+    private final Executor executor;
+    private final HttpClient httpClient;
+    private final Duration maxErrorDuration;
+    private final ScheduledExecutorService errorScheduledExecutor;
+    private final RemoteTaskStats stats;
+    private final boolean isBinaryTransportEnabled;
+
+    @GuardedBy("this")
+    private ListenableFuture<BaseResponse<TaskStatus>> future;
+
+    @Inject
+    public ContinuousBatchTaskStatusFetcher(
+            @ForScheduler HttpClient httpClient,
+            ScheduledExecutorService errorScheduledExecutor,
+            RemoteTaskStats stats,
+            TaskManagerConfig taskConfig,
+            QueryManagerConfig config,
+            InternalCommunicationConfig communicationConfig)
+    {
+        idWorkerMap = new ConcurrentHashMap<>();
+        workerTaskMap = new ConcurrentHashMap<>();
+
+        this.onFail = this::failTask;
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+
+        ExecutorService coreExecutor = newCachedThreadPool(daemonThreadsNamed("remote-task-callback-%s"));
+        this.executor = new BoundedExecutor(coreExecutor, config.getRemoteTaskMaxCallbackThreads());
+
+        Duration refreshMaxWait = taskConfig.getInfoRefreshMaxWait();
+        this.refreshMaxWait = requireNonNull(refreshMaxWait, "refreshMaxWait is null");
+
+        this.maxErrorDuration = config.getRemoteTaskMaxErrorDuration();
+        this.errorScheduledExecutor = requireNonNull(errorScheduledExecutor, "errorScheduledExecutor is null");
+        this.stats = requireNonNull(stats, "stats is null");
+        isBinaryTransportEnabled = requireNonNull(communicationConfig, "communicationConfig is null").isBinaryTransportEnabled();
+    }
+
+    private void failTask(Throwable cause)
+    {
+        // Dummy method for onFail
```

Comment:
let's just put some log statement in case this causes any failures will know when debugging.